core/src/main/java/io/grpc/internal/DelayedClientCall.java

```diff
@@ -141,15 +143,34 @@ public void run() {
    * <p>No-op if either this method or {@link #cancel} have already been called.
    */
   // When this method returns, passThrough is guaranteed to be true
-  public final void setCall(ClientCall<ReqT, RespT> call) {
+  public final Runnable setCall(ClientCall<ReqT, RespT> call) {
+    Listener<RespT> savedListener;
+    Metadata savedHeaders;
     synchronized (this) {
       // If realCall != null, then either setCall() or cancel() has been called.
       if (realCall != null) {
-        return;
+        return null;
       }
       setRealCall(checkNotNull(call, "call"));
+      savedListener = listener;
+      savedHeaders = headers;
+      if (savedListener == null) {
+        assert pendingRunnables.isEmpty();
+        pendingRunnables = null;
+        passThrough = true;
+      }
+    }
+    if (savedListener == null) {
+      return null;
+    } else {
+      realCall.start(savedListener, savedHeaders);
```

Comment:
Why is it important to call start() before returning? For cancel behavior?

Reply:
if `start()` is returned inside `toRun`, then it might cause cancel before start on real stream because there is no guarantee the order. cancel before start on real stream can be dangerous. right?

Reply:
> then it might cause cancel before start on real call because there is no guarantee the order

Handling that sort of thing is the entire point of `DelayedClientCall`, but let's say it didn't handle that case. Then it must be fixed, because the `setCall()` here can run concurrently with the application calling `call.cancel()`. It has nothing to do with "before returning". Instead, we'd have to call `realCall.start()` _within the lock_, or have some other way to protect it.

But I'm confident DelayedClientCall handles that case. If cancel is called before `setCall()`, then the call listener is called immediately and NOOP_CALL is set as the real call. If cancel is called after `setCall()`, then `delayOrExecute()` is used, which preserves ordering.

Reply:
> If cancel is called after setCall(), then delayOrExecute() is used, which preserves ordering.

hm, right, it looks only minimum changes are needed to achieve the original goal that `getExecutor()` is skipped if call is cancelled.